go/parquet/internal/utils/bit_block_counter.go

```diff
@@ -0,0 +1,260 @@
+func (b *BitBlockCounter) getBlockSlow(blockSize int64) BitBlockCount {
+	runlen := int16(Min(b.bitsRemaining, blockSize))
+	popcnt := int16(bitutil.CountSetBits(b.bitmap, int(b.bitOffset), int(runlen)))
+	b.bitsRemaining -= int64(runlen)
+	b.bitmap = b.bitmap[runlen/8:]
+	return BitBlockCount{runlen, popcnt}
+}
+
+// Return the next run of available bits, usually 256. The returned
+// pair contains the size of run and the number of true values. The last
+// block will have a length less than 256 if the bitmap length is not a
+// multiple of 256, and will return 0-length blocks in subsequent
+// invocations.
+func (b *BitBlockCounter) NextFourWords() BitBlockCount {
+	if b.bitsRemaining == 0 {
+		return BitBlockCount{0, 0}
+	}
+
+	totalPopcnt := 0
+	if b.bitOffset == 0 {
+		// if we're aligned at 0 bitoffset, then we can easily just jump from
+		// word to word nice and easy.
+		if b.bitsRemaining < FourWordsBits {
+			return b.getBlockSlow(FourWordsBits)
+		}
+		totalPopcnt += bits.OnesCount64(loadWord(b.bitmap))
+		totalPopcnt += bits.OnesCount64(loadWord(b.bitmap[8:]))
+		totalPopcnt += bits.OnesCount64(loadWord(b.bitmap[16:]))
+		totalPopcnt += bits.OnesCount64(loadWord(b.bitmap[24:]))
+	} else {
+		// When the offset is > 0, we need there to be a word beyond the last
+		// aligned word in the bitmap for the bit shifting logic.
+		if b.bitsRemaining < 5*FourWordsBits-int64(b.bitOffset) {
+			return b.getBlockSlow(FourWordsBits)
+		}
+
+		current := loadWord(b.bitmap)
+		next := loadWord(b.bitmap[8:])
+		totalPopcnt += bits.OnesCount64(shiftWord(current, next, int64(b.bitOffset)))
+
+		current = next
+		next = loadWord(b.bitmap[16:])
+		totalPopcnt += bits.OnesCount64(shiftWord(current, next, int64(b.bitOffset)))
+
+		current = next
+		next = loadWord(b.bitmap[24:])
+		totalPopcnt += bits.OnesCount64(shiftWord(current, next, int64(b.bitOffset)))
+
+		current = next
+		next = loadWord(b.bitmap[32:])
+		totalPopcnt += bits.OnesCount64(shiftWord(current, next, int64(b.bitOffset)))
+	}
+	b.bitmap = b.bitmap[bitutil.BytesForBits(FourWordsBits):]
+	b.bitsRemaining -= FourWordsBits
+	return BitBlockCount{256, int16(totalPopcnt)}
+}
+
+// Return the next run of available bits, usually 64. The returned
+// pair contains the size of run and the number of true values. The last
+// block will have a length less than 64 if the bitmap length is not a
+// multiple of 64, and will return 0-length blocks in subsequent
+// invocations.
+func (b *BitBlockCounter) NextWord() BitBlockCount {
+	if b.bitsRemaining == 0 {
+		return BitBlockCount{0, 0}
+	}
+	popcnt := 0
+	if b.bitOffset == 0 {
+		if b.bitsRemaining < WordBits {
+			return b.getBlockSlow(WordBits)
+		}
+		popcnt = bits.OnesCount64(loadWord(b.bitmap))
+	} else {
+		// When the offset is > 0, we need there to be a word beyond the last
+		// aligned word in the bitmap for the bit shifting logic.
+		if b.bitsRemaining < 2*WordBits-int64(b.bitOffset) {
```

Comment:
I assume no spacing with the multiple is inline with Go style guide? (it reads strangely to me)

Reply:
yea, this is the way it got formatted by running `go fmt`.